File: algorithms/trees/binary-indexed-tree-2d/scala/BinaryIndexedTree2D.scala
object BinaryIndexedTree2D {

  class BIT2D(val rows: Int, val cols: Int) {
    val tree = Array.ofDim[Long](rows + 1, cols + 1)

    def update(r: Int, c: Int, v: Long): Unit = {
      var i = r + 1
      while (i <= rows) {
        var j = c + 1
        while (j <= cols) { tree(i)(j) += v; j += j & (-j) }
        i += i & (-i)
      }
    }

    def query(r: Int, c: Int): Long = {
      var s = 0L; var i = r + 1
      while (i > 0) {
        var j = c + 1
        while (j > 0) { s += tree(i)(j); j -= j & (-j) }
        i -= i & (-i)
      }
      s
    }
  }

  def main(args: Array[String]): Unit = {
    val input = scala.io.StdIn.readLine().trim.split("\\s+").map(_.toInt)
    var idx = 0
    val rows = input(idx); idx += 1; val cols = input(idx); idx += 1
    val bit = new BIT2D(rows, cols)
    for (r <- 0 until rows; c <- 0 until cols) { val v = input(idx); idx += 1; if (v != 0) bit.update(r, c, v) }
    val q = input(idx); idx += 1
    val results = scala.collection.mutable.ArrayBuffer[Long]()
    for (_ <- 0 until q) {
      val t = input(idx); idx += 1; val r = input(idx); idx += 1
      val c = input(idx); idx += 1; val v = input(idx); idx += 1
      if (t == 1) bit.update(r, c, v) else results += bit.query(r, c)
    }
    println(results.mkString(" "))
  }
}

File: algorithms/trees/binary-indexed-tree-2d/csharp/BinaryIndexedTree2D.cs
using System;
using System.Collections.Generic;

public class BinaryIndexedTree2D
{
    long[,] tree;
    int rows, cols;

    public BinaryIndexedTree2D(int rows, int cols)
    {
        this.rows = rows; this.cols = cols;
        tree = new long[rows + 1, cols + 1];
    }

    public void Update(int r, int c, long val)
    {
        for (int i = r + 1; i <= rows; i += i & (-i))
            for (int j = c + 1; j <= cols; j += j & (-j))
                tree[i, j] += val;
    }

    public long Query(int r, int c)
    {
        long s = 0;
        for (int i = r + 1; i > 0; i -= i & (-i))
            for (int j = c + 1; j > 0; j -= j & (-j))
                s += tree[i, j];
        return s;
    }

    public static void Main(string[] args)
    {
        var tokens = Console.ReadLine().Trim().Split();
        int idx = 0;
        int rows = int.Parse(tokens[idx++]), cols = int.Parse(tokens[idx++]);
        var bit = new BinaryIndexedTree2D(rows, cols);
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
            {
                int v = int.Parse(tokens[idx++]);
                if (v != 0) bit.Update(r, c, v);
            }
        int q = int.Parse(tokens[idx++]);
        var results = new List<string>();
        for (int i = 0; i < q; i++)
        {
            int t = int.Parse(tokens[idx++]), r = int.Parse(tokens[idx++]);
            int c = int.Parse(tokens[idx++]), v = int.Parse(tokens[idx++]);
            if (t == 1) bit.Update(r, c, v);
            else results.Add(bit.Query(r, c).ToString());
        }
        Console.WriteLine(string.Join(" ", results));
    }
}
